File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
class Solution {
    public Node connect(Node root) {
        if (root == null || root.right == null) {
            return root;
        }

        Queue<Node> que = new ArrayDeque<>();
        que.add(root.right);
        que.add(root.left);

        while (!que.isEmpty()) {
            int n = que.size();
            Node before = null;

            for (int i = 0; i < n; i++) {
                Node now = que.poll();
                if (i != 0) {
                    now.next = before;
                }
                if (now.right != null) {
                    que.add(now.right);
                    que.add(now.left);
                }
                before = now;
            }
        }

        return root;
    }
}

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.py
from collections import deque


class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root or not root.left:
            return root

        que = deque()
        que.append(root.right)
        que.append(root.left)

        while que:
            n = len(que)
            before = None
            for i in range(n):
                now = que.popleft()
                if i:
                    now.next = before
                if now.right and now.left:
                    que.append(now.right)
                    que.append(now.left)
                before = now

        return root